library/math/dual.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; dual.lisp
+
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/classes
+   #:coalton-library/functions
+   #:coalton-library/math/arith
+   #:coalton-library/math/elementary
+   #:coalton-library/math/integral
+   #:coalton-library/hash)
+  (:export
+   #:Dual
+   #:primal-part
+   #:dual-part)
+  (:documentation "
+dual numbers are a hypercomplex number system [1]. A dual number has the form
+a + bε where a and b are real numbers and ε is a symbol that satisfies ε^2=0
+and ε!=0. One application of dual numbers is automatic differentiation; an example
+taken from [2] is as follows:
+
+    consider you have the given expression f(x) = 3x+2 and you want to calculate
+    f(4) and f'(4).
+
+    You then convert 4 into a dual number which gives:
+
+        4+1ε
+
+    then the computation is as follows:
+
+        (4*1ε) * (3 + 0ε) =
+        12 * 0ε + 3ε + 0ε^2 =
+        12 + 3ε
+
+    lastly, you need to add the constant 2 of the expression
+
+        (12 + 3ε) + (2 + 0ε) =
+        14 + 3ε
+
+   in this result, the primal 14 is the value of f(4) and the dual is the value of
+   of f'(4).
+
+   Haskell has an automatic differentiation library and you can find it here
+   https://hackage.haskell.org/package/ad.
+
+   limitations:
+
+     some limitations is that `Ord` and `Eq` and `Hash` only work on primals.
+     but the advantages of this approach is that code will be auto differentiated
+     for free.
+
+   references:
+
+   [1] https://en.wikipedia.org/wiki/Dual_number
+   [2] https://blog.demofox.org/2014/12/30/dual-numbers-automatic-differentiation/"))
+
+(in-package #:coalton-library/math/dual)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+(coalton-toplevel
+
+  (define-type (Dual :t)
+    "Representation of a Dual number in the form (a + bε) where a and b are  real numbers and ε satisfie ε**2 = 0 and ε != 0.  Note: `Eq`, and `Ord` and `Hash` only work on the primal component."
```

Comment:
Some more typographic mistakes. I don't want to hold up this PR, but these should be fixed in a follow-up PR.

- lowercase "Dual" in the description
- extra unnecessary spaces between "are real"
- no need for parens around $a+b\varepsilon$
- misspelling "satisfie"
- inconsistent use of `^` and `**`. We recommend using `^` to be consistent with our standard library.
- "only work on" -> "only make use of"